crates/exex/src/serde.rs

```diff
@@ -2326,4 +2419,107 @@ mod tests {
+            identifier.full_name,
+            Some("starkware.cairo.common.cairo_builtins.KeccakBuiltin".to_string())
+        );
+        assert_eq!(identifier.type_, Some("struct".to_string()));
+    }
+
+    #[test]
+    fn test_get_struct_definition_no_match() {
+        let program = setup_program(&TestProgram::ModelOption);
+        let result = get_struct_definition(&program, "NonExistentStruct");
+
+        // Assert that the result is an error indicating a type mismatch directly
+        assert_eq!(
+            result,
+            Err(KakarotSerdeError::StructNotFound("NonExistentStruct".to_string(), 0))
+        );
+    }
+
+    #[test]
+    fn test_get_struct_definition_multiple_matches() {
+        let program = setup_program(&TestProgram::ModelMemory);
+        let result = get_struct_definition(&program, "ImplicitArgs");
+
+        // Assert that the result is an error indicating multiple matches
+        assert_eq!(result, Err(KakarotSerdeError::StructNotFound("ImplicitArgs".to_string(), 3)));
+    }
+
+    #[test]
+    fn test_serialize_struct_none_pointer() {
+        let kakarot_serde = setup_kakarot_serde(&TestProgram::KeccakAddUint256);
+
+        // Test with a None pointer, expecting an empty vector as the result
+        let result = kakarot_serde.serialize_struct("SomeStruct", None);
+        assert_eq!(result, Ok(HashMap::new()));
+    }
+
+    #[test]
+    fn test_serialize_struct_missing_struct_definition() {
+        let kakarot_serde = setup_kakarot_serde(&TestProgram::KeccakAddUint256);
+
+        // Test with a struct name that doesn't exist in the program
+        let ptr = Some(Relocatable { segment_index: 0, offset: 0 });
+        let result = kakarot_serde.serialize_struct("NonExistentStruct", ptr);
+
+        // Assert that the error indicates the struct was not found
+        assert_eq!(
+            result,
+            Err(KakarotSerdeError::StructNotFound("NonExistentStruct".to_string(), 0))
+        );
+    }
+
+    #[test]
+    fn test_serialize_struct_valid_struct() {
+        let mut kakarot_serde = setup_kakarot_serde(&TestProgram::KeccakAddUint256);
+
+        // Setup the struct members
+        // We need to insert the struct members into memory at:
+        // - Offset 0
+        // - Offset 8
+        let member_value1 = MaybeRelocatable::Int(Felt252::from(42));
+        let member_value2 = MaybeRelocatable::Int(Felt252::from(93));
+        let ptr = kakarot_serde
+            .runner
+            .vm
+            .gen_arg(&vec![
+                member_value1.clone(),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                MaybeRelocatable::Int(Felt252::ZERO),
+                member_value2.clone(),
```

Comment:
this is not a valid KeccackBuiltin struct, which is defined as

```
struct KeccakBuiltin {
    input: KeccakBuiltinState,
    output: KeccakBuiltinState,
}
```

with

```
struct KeccakBuiltinState {
    s0: felt,
    s1: felt,
    s2: felt,
    s3: felt,
    s4: felt,
    s5: felt,
    s6: felt,
    s7: felt,
}
```

meaning that `KeccakBuiltin` is 2 x 8 = 16 keys long

Reply:
Yes you mean that my member_value1 and member_value2 should be Relocatable pointing to my two KeccakBuiltinState respectively right?

But here this is not supposed to be realistic, just wanted to test with the minimal setup to ensure that we fetch properly the values. So for this test and only for this test (and not in real life):

```rust
 struct KeccakBuiltin {
    input: felt,
    output: felt,
}
```

Reply:
And even if I wanted to, right now in my `serialize_inner` I don't have the proper tools needed to serialize it completely (only have felt stuffs)

Reply:
no no, not relocatable, in cairo zero, this definition

```
struct KeccakBuiltin {
    input: KeccakBuiltinState,
    output: KeccakBuiltinState,
}
```

actually inlines the `KeccakBuiltinState`, and so lead to the exact same memory layout as (leaving the conflicting name for the sake of clarity)

```
struct KeccakBuiltin {
    s0: felt,
    s1: felt,
    s2: felt,
    s3: felt,
    s4: felt,
    s5: felt,
    s6: felt,
    s7: felt,
    s0: felt,
    s1: felt,
    s2: felt,
    s3: felt,
    s4: felt,
    s5: felt,
    s6: felt,
    s7: felt,
}
```